src/pytkdocs/parsers/docstrings/__init__.py

```diff
@@ -0,0 +1,122 @@
+
+
+class AnnotatedObject:
+    """A helper class to store information about an annotated object."""
+
+    def __init__(self, annotation, description):
+        self.annotation = annotation
+        self.description = description
+
+
+class Parameter(AnnotatedObject):
+    """A helper class to store information about a signature parameter."""
+
+    def __init__(self, name, annotation, description, kind, default=empty):
+        super().__init__(annotation, description)
+        self.name = name
+        self.kind = kind
+        self.default = default
+
+    def __str__(self):
+        return self.name
+
+    def __repr__(self):
+        return f"<Parameter({self.name}, {self.annotation}, {self.description}, {self.kind}, {self.default})>"
+
+    @property
+    def is_optional(self):
+        return self.default is not empty
+
+    @property
+    def is_required(self):
+        return not self.is_optional
+
+    @property
+    def is_args(self):
+        return self.kind is inspect.Parameter.VAR_POSITIONAL
+
+    @property
+    def is_kwargs(self):
+        return self.kind is inspect.Parameter.VAR_KEYWORD
+
+    @property
+    def default_string(self):
+        if self.is_kwargs:
+            return "{}"
+        elif self.is_args:
+            return "()"
+        elif self.is_required:
+            return ""
+        return repr(self.default)
+
+
+class Section:
+    """A helper class to store a docstring section."""
+
+    class Type:
+        MARKDOWN = "markdown"
+        PARAMETERS = "parameters"
+        EXCEPTIONS = "exceptions"
+        RETURN = "return"
+
+    def __init__(self, section_type, value):
+        self.type = section_type
+        self.value = value
+
+    def __str__(self):
+        return self.type
+
+    def __repr__(self):
+        return f"<Section(type={self.type!r})>"
+
+
+class Parser(metaclass=ABCMeta):
+    """
+    A class to parse docstrings.
+
+    It is instantiated with an object's path, docstring, signature and return type.
+
+    The `parse` method then returns structured data,
+    in the form of a list of [`Section`][pytkdocs.parsers.docstrings.Section]s.
+    It also return the list of errors that occurred during parsing.
+    """
+
+    def __init__(
+        self,
+        path: str,
+        docstring: str,
+        signature: Optional[inspect.Signature] = None,
+        return_type: Optional[Any] = empty,
+    ) -> None:
+        """
+        Arguments:
+            path: An object's dotted-path, used to improve error messages.
+            docstring: An object's docstring: the docstring to parse.
+            signature: An object's signature if any.
+            return_type: An object's return type if any. Can be a string or a type.
+        """
+        self.path = path
+        self.docstring = docstring or ""
+        self.signature = signature
+        self.return_type = return_type
+        self.parsing_errors: List[str] = []
+
+    def do_parse(self, **kwargs: Any):
```

Comment:
What is the return type? `Tuple[List[Section],List[Error]]` ?

Reply:
Just `List[Section]`, the tuple is returned by `parse`.

Reply:
Would be good to have these annotations on the return types as well for both methods.

Reply:
Sure, I simply forgot them hehe